File: Source Code/Transmitter/src/com/dvt/kinderfinder/transmitter/MainActivity.java
package com.dvt.kinderfinder.transmitter;

import java.util.LinkedList;

import org.apache.http.HttpStatus;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

public class MainActivity extends ActionBarActivity {
	private Spinner restaurantList;
	private Spinner typeList;
	private Button transmitButton;
	private EditText xPosBox, yPosBox;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		restaurantList = (Spinner) findViewById(R.id.restaurantList);
		typeList = (Spinner) findViewById(R.id.typeList);
		xPosBox = (EditText) findViewById(R.id.xPosBox);
		yPosBox = (EditText) findViewById(R.id.yPosBox);
		transmitButton = (Button) findViewById(R.id.transmitButton);
		
		restaurantList.setOnItemSelectedListener(spinnerItemSelectedhandler);
		transmitButton.setOnClickListener(buttonClickHandler);
		
		xPosBox.clearFocus();
		loadRestaurants();
	}
	
	/**
	 * When an item in the restaurant list is selected, loads the list of tags
	 * that belong to the restaurant.
	 */
	private OnItemSelectedListener spinnerItemSelectedhandler = new OnItemSelectedListener() {

		@Override
		public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
			loadTypes(restaurantList.getSelectedItem().toString());
		}

		@Override
		public void onNothingSelected(AdapterView<?> parent) { }
	};
	
	private OnClickListener buttonClickHandler = new OnClickListener() {
		@Override
		public void onClick(View v) {
			String restaurant = restaurantList.getSelectedItem().toString();
			String type = typeList.getSelectedItem().toString();
			String x = xPosBox.getText().toString();
			String y = yPosBox.getText().toString();
			
			if (restaurant.equals("") || type.equals("")) {
				Toast.makeText(getApplicationContext(), "Something went wrong; please try again", Toast.LENGTH_SHORT).show();
				return;
			}
			else if (x.equals("") || y.equals("")) {
				Toast.makeText(getApplicationContext(), "Please enter x and y co-ords", Toast.LENGTH_SHORT).show();
				return;
			}
			
			JsonBuilder jb = new JsonBuilder();
			jb.addEntry("RestaurantName", restaurant);
			jb.addEntry("TransmitterType", type);
			jb.addEntry("X", x);
			jb.addEntry("Y", y);
			
			new RequestTask() {
				@Override
				protected void onPostExecute(String result) {
					if (statusCode == HttpStatus.SC_OK) {
						Intent intent = new Intent(getApplicationContext(), TransmitActivity.class);
						intent.putExtra("ID", result);
						startActivity(intent);
						finish();
					}
					else {
						Toast.makeText(getApplicationContext(), "An error occured; please try again", Toast.LENGTH_SHORT).show();
					}
				}
			}.execute("api/assignrestaurant", jb.toString());
		}
	};
	
	/**
	 * Requests a list of all restaurants from the server.
	 */
	private void loadRestaurants() {
		restaurantList.setEnabled(false);
		typeList.setEnabled(false);
		xPosBox.setEnabled(false);
		yPosBox.setEnabled(false);
		transmitButton.setEnabled(false);
		
		new RequestTask() {
			@Override
			protected void onPostExecute(String result) {
				if (statusCode == HttpStatus.SC_OK) {
					LinkedList<String> restaurants = JsonBuilder.JsonToList(result);
					ArrayAdapter<String> adapter = new ArrayAdapter<>(getApplicationContext(), R.layout.spinner_item, restaurants);
					adapter.setDropDownViewResource(R.layout.spinner_dropdown_item);
					restaurantList.setAdapter(adapter);
					
					if (restaurants.size() > 0) {
						restaurantList.setEnabled(true);
					}
				}
				else {
					Toast.makeText(getApplicationContext(), "Error contacting server", Toast.LENGTH_SHORT).show();
				}
			}
		}.execute("api/restaurantlist");
	}
	
	/**
	 * Requests a list of tags that belong to a specific restaurant from the
	 * server.
	 */
	private void loadTypes(String restaurant) {
		JsonBuilder jb = new JsonBuilder();
		jb.addEntry("RestaurantName", restaurant);
		
		new RequestTask() {
			@Override
			protected void onPostExecute(String result) {
				if (statusCode == HttpStatus.SC_OK) {
					LinkedList<String> types = JsonBuilder.JsonToList(result);
					ArrayAdapter<String> adapter = new ArrayAdapter<>(getApplicationContext(), R.layout.spinner_item, types);
					adapter.setDropDownViewResource(R.layout.spinner_dropdown_item);
					typeList.setAdapter(adapter);
					
					if (types.size() > 0) {
						typeList.setEnabled(true);
						xPosBox.setEnabled(true);
						yPosBox.setEnabled(true);
						transmitButton.setEnabled(true);
					}
					else {
						Toast.makeText(getApplicationContext(), "This restaurant has no more available types", Toast.LENGTH_LONG).show();
					}
				}
				else {
					Toast.makeText(getApplicationContext(), "Error contacting server", Toast.LENGTH_SHORT).show();
				}
			}
		}.execute("api/transmittertype", jb.toString());
	}
}

File: Source Code/KinderFinder/Transmitter/MainActivity.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Threading;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Widget;

using Transmitter.Utility;

namespace Transmitter {

	[Activity(Label = "Transmitter", MainLauncher = true, Icon = "@drawable/icon")]
	public class MainActivity : Activity {
		ISharedPreferences Pref;
		ISharedPreferencesEditor Editor;

		Spinner RestaurantList, TypeList;
		EditText XPosBox, YPosBox;
		Button RefreshButton, StartButton, StopButton;
		ProgressBar ProgressBar;

		protected override void OnCreate(Bundle bundle) {
			base.OnCreate(bundle);
			SetContentView(Resource.Layout.Main);

			Pref = GetSharedPreferences("Preferences", 0);
			Editor = Pref.Edit();

			RestaurantList = FindViewById<Spinner>(Resource.Id.RestaurantList);
			TypeList = FindViewById<Spinner>(Resource.Id.TypeList);
			XPosBox = FindViewById<EditText>(Resource.Id.XPos);
			YPosBox = FindViewById<EditText>(Resource.Id.YPos);
			RefreshButton = FindViewById<Button>(Resource.Id.Refresh);
			StartButton = FindViewById<Button>(Resource.Id.Start);
			StopButton = FindViewById<Button>(Resource.Id.Stop);
			ProgressBar = FindViewById<ProgressBar>(Resource.Id.ProgressBar);

			RestaurantList.ItemSelected += (sender, e) => LoadTypes(RestaurantList.SelectedItem.ToString());
			RefreshButton.Click += RefreshPressed;
			StartButton.Click += StartPressed;
			StopButton.Click += StopPressed;

			CheckService();
		}

		void RefreshPressed(object sender, EventArgs e) {
			CheckService();
		}

		void CheckService() {
			if (TransmitService.IsRunning()) {
				RestaurantList.Enabled = false;
				TypeList.Enabled = false;
				XPosBox.Enabled = false;
				YPosBox.Enabled = false;
				StartButton.Enabled = false;
				StopButton.Enabled = true;
				ProgressBar.Visibility = Android.Views.ViewStates.Invisible;
			}
			else {
				RestaurantList.Enabled = false;
				TypeList.Enabled = false;
				XPosBox.Enabled = false;
				YPosBox.Enabled = false;
				StartButton.Enabled = false;
				StopButton.Enabled = false;
				LoadRestaurants();
			}
		}

		void LoadRestaurants() {
			ProgressBar.Visibility = Android.Views.ViewStates.Visible;

			ThreadPool.QueueUserWorkItem(state => {
				var reply = AppTools.SendRequest("api/restaurantlist", null);

				switch (reply.StatusCode) {
					case HttpStatusCode.OK:
						var list = Deserialiser<List<string>>.Run(reply.Body);
						var adapter = new ArrayAdapter<string>(this, Android.Resource.Layout.SimpleSpinnerItem, list);
						adapter.SetDropDownViewResource(Android.Resource.Layout.SimpleSpinnerDropDownItem);

						RunOnUiThread(() => {
							RestaurantList.Adapter = adapter;
							RestaurantList.Enabled = list.Count > 0;
						});

						break;
					default:
						RunOnUiThread(() => Toast.MakeText(this, "Error contacting server: " + reply.StatusCode, ToastLength.Short).Show());
						break;
				}

				RunOnUiThread(() => ProgressBar.Visibility = Android.Views.ViewStates.Invisible);
			});
		}

		void LoadTypes(string restaurant) {
			ProgressBar.Visibility = Android.Views.ViewStates.Visible;
			var jb = new JsonBuilder();
			jb.AddEntry("RestaurantName", restaurant);

			ThreadPool.QueueUserWorkItem(state => {
				var reply = AppTools.SendRequest("api/transmittertype", jb.ToString());

				switch (reply.StatusCode) {
					case HttpStatusCode.OK:
						var list = Deserialiser<List<string>>.Run(reply.Body);
						var adapter = new ArrayAdapter<string>(this, Android.Resource.Layout.SimpleSpinnerItem, list);
						adapter.SetDropDownViewResource(Android.Resource.Layout.SimpleSpinnerDropDownItem);

						RunOnUiThread(() => {
							TypeList.Adapter = adapter;
							bool enable = list.Count > 0;
							TypeList.Enabled = enable;
							XPosBox.Enabled = enable;
							YPosBox.Enabled = enable;
							StartButton.Enabled = enable;
						});

						break;
					case HttpStatusCode.BadRequest:
						RunOnUiThread(() => Toast.MakeText(this, "Transmitter ID not accepted; restart app", ToastLength.Short).Show());
						break;
					case HttpStatusCode.Conflict:
						RunOnUiThread(() => Toast.MakeText(this, "No more available transmitter types", ToastLength.Short).Show());
						break;
					default:
						RunOnUiThread(() => Toast.MakeText(this, "Error contacting server: " + reply.StatusCode, ToastLength.Short).Show());
						break;
				}

				RunOnUiThread(() => ProgressBar.Visibility = Android.Views.ViewStates.Invisible);
			});
		}

		void StartPressed(object sender, EventArgs e) {
			string restaurant = RestaurantList.SelectedItem.ToString();
			string type = TypeList.SelectedItem.ToString();
			string xPos = XPosBox.Text;
			string yPos = YPosBox.Text;

			if (restaurant == "" || type == "") {
				Toast.MakeText(this, "Something went wrong; please try again", ToastLength.Short).Show();
				return;
			}
			else if (xPos == "" || yPos == "") {
				Toast.MakeText(this, "Please enter x and y co-ords", ToastLength.Short).Show();
				return;
			}

			var jb = new JsonBuilder();
			jb.AddEntry("RestaurantName", restaurant);
			jb.AddEntry("TransmitterType", type);
			jb.AddEntry("X", xPos);
			jb.AddEntry("Y", yPos);

			ThreadPool.QueueUserWorkItem(state => {
				var reply = AppTools.SendRequest("api/assignrestaurant", jb.ToString());

				switch (reply.StatusCode) {
					case HttpStatusCode.OK:
						Editor.PutString("ID", reply.Body);
						Editor.Commit();
						StartService(new Intent(this, typeof(TransmitService)));
						RunOnUiThread(CheckService);
						break;
					case HttpStatusCode.BadRequest:
						RunOnUiThread(() => Toast.MakeText(this, "Transmitter ID not accepted; restart app", ToastLength.Short).Show());
						break;
					case HttpStatusCode.Conflict:
						RunOnUiThread(() => Toast.MakeText(this, "Transmitter type in use; press refresh", ToastLength.Short).Show());
						break;
					default:
						RunOnUiThread(() => Toast.MakeText(this, "Error contacting server: " + reply.StatusCode, ToastLength.Short).Show());
						break;
				}
			});
		}

		void StopPressed(object sender, EventArgs e) {
			// Needed to add a slight delay to calling CheckService(), otherwise the service isn't stopped yet so it
			// thinks the service is still running.
			StopService(new Intent(this, typeof(TransmitService)));
			ProgressBar.Visibility = Android.Views.ViewStates.Visible;

			new Thread(() => {
				Thread.Sleep(100);
				RunOnUiThread(() => {
					CheckService();
					ProgressBar.Visibility = Android.Views.ViewStates.Visible;
				});
			}).Start();
		}
	}
}
